doc/compiler/il/GlRegDeps.md

```diff
@@ -0,0 +1,152 @@
+<!--
+Copyright (c) 2021, 2021 IBM Corp. and others
+
+This program and the accompanying materials are made available under
+the terms of the Eclipse Public License 2.0 which accompanies this
+distribution and is available at https://www.eclipse.org/legal/epl-2.0/
+or the Apache License, Version 2.0 which accompanies this distribution and
+is available at https://www.apache.org/licenses/LICENSE-2.0.
+
+This Source Code may also be made available under the following
+Secondary Licenses when the conditions for such availability set
+forth in the Eclipse Public License, v. 2.0 are satisfied: GNU
+General Public License, version 2 with the GNU Classpath
+Exception [1] and GNU General Public License, version 2 with the
+OpenJDK Assembly Exception [2].
+
+[1] https://www.gnu.org/software/classpath/license.html
+[2] http://openjdk.java.net/legal/assembly-exception.html
+
+SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+-->
+
+# Global Register Dependency
+
+`GlRegDeps` is an IL level construct that models code generator register dependencies. It associates virtual registers with real registers, or pairs virtual registers with real registers.
+
+Register dependencies enforce constraints. They represent that a particular value (IL node) has to be in a particular real register at a particular point in the program. These particular points in the program are usually entry or exit corresponding to basic blocks or extended basic blocks where there is control flow. Fundamentally we can think of global register dependencies as a way of conveying information within real (hardware) registers across block boundaries.
+
+Most common control flow opcodes are: `if`, `goto`, `return`, `athrow`, `switch`. They should be the last tree before the `BBEnd`. They should have their own `GlRegDeps` for the branch target if required. The `GlRegDeps` under `BBEnd` is for the fall through case. It is important to make sure the right value is placed in the right real register at the point where the `GlRegDeps` is evaluated.
+
+If a value that is expected to be in a register is not in the register, or the register is not preserved if it is a call, it leads to register shuffling. The local register allocator has to shuffle it to a different register to make sure it has the right value in `GlRegDeps`. Register shuffling does come at a cost. It should be avoided when possible.
+
+Global Register Allocation (`GRA`) related opcodes such as `GlRegDeps` use reference counts to determine when a register is no longer needed. It is a basic rule within the codegen for nodes and their associated virtual registers in general. It has to be set up correctly, otherwise it could result in registers being overwritten with some other values prematurely. In the following example, `n190n` is referenced `5` times but its reference count is set up incorrectly as `6` (`rc=6`).
```

Comment:
I'm guessing that if the reference count is too low, a register might be overwritten prematurely, but if it's too great, the register might be held too long.  Is that correct, or are my guesses at the consequences completely wrong?

Reply:
That's my understanding too. Also elaborated in https://github.com/eclipse/omr/pull/5898#discussion_r608084005

Reply:
Sorry for the duplicate comment.  I think I was reviewing when you did a "force push", and it looked like my original comment had disappeared.